### ISSUE TITLE
Changes in input file properties should not cause incremental build

### ISSUE THREAD

**Issue by hungvietnguyen** [gradle/gradle#4114]
### Expected Behavior
Gradle correctly reports out-of-date and removed files in incremental builds.

### Current Behavior
Gradle does not report any out-of-date or removed files in incremental builds when task inputs are configured with `task.getInputs().dir(...)`.

### Context
I want to configure task inputs to enable incremental build. Since the number of input directories may change across builds, I need to configure them dynamically using `task.getInputs().dir(...)`.

### Steps to Reproduce
See the attached project for details: [PathSensitivtyBug.zip](https://github.com/gradle/gradle/files/1653236/PathSensitivtyBug.zip)

The project has the following structure:
- `test1/0`: input directory for the first build
- `test2/1` and `test2/2`: input directories for the second build
- `build.gradle`: the build file

Step 1: Run `gradlew sampleTask -Ptest=1`
Step 2: Run `gradlew sampleTask -Ptest=2`. The console shows "Task is incremental" but does not show any [OUT-OF-DATE] or [REMOVED] input files. This is not expected.

### Your Environment
Gradle version: 4.4

### Additional Request on PathSensitivity Documentation
`PathSensitivity` seems to be very lightly documented. In particular, the semantics of `PathSensitivity.RELATIVE` when used with `@InputFile` and `@InputDirectory` seem to be different but are not clearly defined.

Specifically, the term "hierarchy" used at https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/PathSensitivity.html is crucial for understanding, but is also vague: What is the root path that the path in question is relative to? When a `PathSensitivity.RELATIVE` is annotated on an `@InputDirectory`, is it applied to the directory itself, or to the files within the directory, or both?

It would be nice to have more docs explaining all of these in more detail.

**Comment by wolfs** [comment]
@hungvietnguyen Thank you for reporting!

I think `PathSensitivity.RELATIVE` doesn't make a difference for exposing this bug, does it? I observed the same behaviour if no path sensitivity is specified.

I think the right behaviour would be that the task is not incremental on the second run, since the number of input file properties changed. Would that be what you expect?

**Comment by hungvietnguyen** [comment]
(1) You are right, I just checked again and also found that the presence of path sensitivity does not make a difference. (I was experimenting with `PathSensitivity.RELATIVE` earlier and didn't expect that this was still a bug without it, but turns out it is.)

(2) Yes that would work. The alternative is that Gradle runs incrementally with the correctly reported changed files/directories. I think both are acceptable, the latter is actually more efficient, but maybe the former is more straightforward.

Note that the above bug is exposed when the number of input directories changes across builds.

However, even when the number of input directories doesn't change across builds, there seem to be a few unexpected details too. I'm listing them here in case you also want to address:

Scenario 1:
```
if (project.property('test').equals("1")) {
    sampleTask.getInputs().dir("test1/0")
} else {
    sampleTask.getInputs().dir("test2/1")
}
```
- Run 1: `gradlew sampleTask -Ptest=1` => Not expected: `test1/0` directory is not reported as out-of-date. (Or at least it is inconsistent with the report when `files()` is used instead of `dir()`, as in scenario 2 below.)
- Run 2: `gradlew sampleTask -Ptest=2` => Not expected: `test1/0` directory is not reported as removed, `test2/1` directory is not reported as changed.

Scenario 2: 
```
if (project.property('test').equals("1")) {
    sampleTask.getInputs().files("test1/0")
} else {
    sampleTask.getInputs().files("test2/1")
}
```
- Run 1: `gradlew sampleTask -Ptest=1` => Not expected: `test1/0/classes.dex `and `test1/0/classes2.dex` are not reported as out-of-date.
- Run 2: `gradlew sampleTask -Ptest=2 `=> Expected (all directories and files are reported).

More scenarios: If you try using a mixture of `dir()` and `files()` with more input directories, then you'll get a mixture of reported directories, missing directories, reported files, missing files.

**Comment by wolfs** [comment]
@hungvietnguyen I presume the new scenarios use relative path sensitivity?

**Comment by hungvietnguyen** [comment]
@wolfs To clarify, the new scenarios above also do not use any path sensitivity.

**Comment by wolfs** [comment]
@hungvietnguyen I opened https://github.com/gradle/gradle/issues/4166 for Scenario 2 which I find surprising.

With respect to scenario 1: The root directories are not considered inputs for input directories. This is why they are not reported as removed, added, etc.

Does this make sense to you?

**Comment by hungvietnguyen** [comment]
@wolfs This makes sense after I read the documentation at https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskInputs.html#dir-java.lang.Object-: "All files found under the given directory are treated as input files for this task." However, this is a little surprising and not intuitive for users of Gradle when the paths to root directories are not considered as inputs. With `@InputFile` and `@InputFiles`, the paths to the given files/directories are considered as inputs, so why is it that `@InputDirectory` suddenly behaves differently?

I assume this must have been a conscious decision from Gradle (I can kind of guess the motivation), but keep in mind that it is not intuitive for users.

Actually, if Gradle decides to change the behavior of `@InputDirectory` to also include the directory's path, then most existing behavior still remains, except for the case of empty input directories: Changing from an empty input directory to another empty input directory used to not trigger an OUT-OF-DATE status, but will now trigger OUT-OF-DATE (which seems Okay).

**Comment by wolfs** [comment]
Fixed via #4137.

**Comment by wolfs** [comment]
@hungvietnguyen I opened https://github.com/gradle/gradle/issues/4204 to track the inconsistency between `@InputDirectory` and `@InputFiles`.
